athenz/resource_role.go

```diff
@@ -169,24 +201,76 @@ func resourceRoleRead(ctx context.Context, d *schema.ResourceData, meta interfac
+	for _, m := range members {
+		if string(member.MemberName) == string(m.MemberName) {
+			return true
+		}
+	}
+	return false
+}
+
+// return all role members that appears in list1 but not includes in list2
+func unifyMembers(list1, list2 []*zms.RoleMember) []*zms.RoleMember {
+	toReturn := make([]*zms.RoleMember, 0)
+	for _, m := range list1 {
+		if !isMembersIncludes(m, list2) {
+			toReturn = append(toReturn, m)
+		}
+	}
+	return toReturn
+}
+
+/*
+since we not allow configuring both members (deprecated) and member attributes at the same time, state changes CAN'T be one of the following:
+1. remove members from both attributes member and members.
+2. add members in both attributes member and members.
+*/
+func handleMembersChange(removeDeprecatedRoleMembers, addDeprecatedRoleMembers, removeRoleMembers, addRoleMembers []*zms.RoleMember) ([]*zms.RoleMember, []*zms.RoleMember) {
+	var removeMembers []*zms.RoleMember
+	if len(removeDeprecatedRoleMembers) == 0 {
+		removeMembers = unifyMembers(removeRoleMembers, addDeprecatedRoleMembers)
+	} else {
+		removeMembers = unifyMembers(removeDeprecatedRoleMembers, addRoleMembers)
+	}
+	if len(addDeprecatedRoleMembers) == 0 {
+		return removeMembers, addRoleMembers
+	}
+	return removeMembers, addDeprecatedRoleMembers
+}
+
 func resourceRoleUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	zmsClient := meta.(client.ZmsClient)
 	dn, rn, err := splitRoleId(d.Id())
 	if err != nil {
 		return diag.FromErr(err)
 	}
 	auditRef := d.Get("audit_ref").(string)
+	removeDeprecatedRoleMembers := make([]*zms.RoleMember, 0)
+	addDeprecatedRoleMembers := make([]*zms.RoleMember, 0)
 	if d.HasChange("members") {
 		if _, ok := d.GetOk("trust"); ok {
 			return diag.Errorf("delegated roles cannot change members")
 		}
 		os, ns := handleChange(d, "members")
-		remove := expandRoleMembers(os.Difference(ns).List())
-		add := expandRoleMembers(ns.Difference(os).List())
-		err := updateRoleMembers(dn, rn, remove, add, auditRef, zmsClient)
-		if err != nil {
-			return diag.Errorf("error updating group membership: %s", err)
+		removeDeprecatedRoleMembers = expandDeprecatedRoleMembers(os.Difference(ns).List())
+		addDeprecatedRoleMembers = expandDeprecatedRoleMembers(ns.Difference(os).List())
+	}
+	removeRoleMembers := make([]*zms.RoleMember, 0)
+	addRoleMembers := make([]*zms.RoleMember, 0)
+	if d.HasChange("member") {
+		if _, ok := d.GetOk("trust"); ok {
+			return diag.Errorf("delegated roles cannot change members")
 		}
+		os, ns := handleChange(d, "member")
+		removeRoleMembers = expandRoleMembers(os.Difference(ns).List())
+		addRoleMembers = expandRoleMembers(ns.Difference(os).List())
+	}
+	removeMembers, addMembers := handleMembersChange(removeDeprecatedRoleMembers, addDeprecatedRoleMembers, removeRoleMembers, addRoleMembers)
+	err = updateRoleMembers(dn, rn, removeMembers, addMembers, auditRef, zmsClient)
```

Comment:
It is too complicated (and might lead to unexpected behavior).
I would suggest to split it to 2 different methods - `membersToAdd` membersToDelete`